docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -102,6 +102,289 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
+                    .filter(partition -> partitionMatches(partitionColumns, effectivePredicate, predicate, partition))
+                    .collect(toImmutableList());
+        }
+        else {
+            List<String> partitionNamesList = hiveTableHandle.getPartitionNames()
+                    .orElseGet(() -> getFilteredPartitionNames(metastore, tableName, partitionColumns, compactEffectivePredicate));
+            partitionsIterable = () -> partitionNamesList.stream()
+                    // Apply extra filters which could not be done by getFilteredPartitionNames
+                    .map(partitionName -> parseValuesAndFilterPartition(tableName, partitionName, partitionColumns, partitionTypes, effectivePredicate, predicate))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .iterator();
+            partitionNames = Optional.of(partitionNamesList);
+        }
+
+        return new HivePartitionResult(partitionColumns, partitionNames, partitionsIterable, effectivePredicate, compactEffectivePredicate, hiveBucketHandle, bucketFilter);
+    }
+
+First, the ``TupleDomain`` from the constraint is intersected with the
+``TupleDomain`` already applied to the ``TableHandle`` to form an
+``effectivePredicate``. If the result is not empty, it can be used as
+a filter on Hive buckets and partitions. Note that we're using ``predicate``
+only for partition filtering.
+
+Now let's take a look at how a new ``HiveTableHandle`` is created from
+an old one and an additional constraint:
+
+.. code-block:: java
+
+    public HiveTableHandle applyPartitionResult(HiveTableHandle handle, HivePartitionResult partitions, Constraint constraint)
+    {
+        Optional<List<String>> partitionNames = partitions.getPartitionNames();
+        Optional<List<HivePartition>> partitionList = Optional.empty();
+        TupleDomain<ColumnHandle> enforcedConstraint = handle.getEnforcedConstraint();
+
+        // Partitions will be loaded if
+        // 1. Number of partitionNames is less than or equal to threshold value. Thereby generating additional filter criteria
+        //    that can be applied on other join side (if the join is based on partition column),
+        // 2. If additional predicate is passed as a part of Constraint. (specified via loadPartition). This delays the partition checks
+        //    until we have additional filtering based on Constraint
+        if (canPartitionsBeLoaded(partitions) || constraint.predicate().isPresent()) {
+            partitionNames = Optional.empty();
+            partitionList = Optional.of(getPartitionsAsList(partitions));
+            List<HiveColumnHandle> partitionColumns = partitions.getPartitionColumns();
+            enforcedConstraint = partitions.getEffectivePredicate().filter((column, domain) -> partitionColumns.contains(column));
+        }
+        return new HiveTableHandle(
```

Comment:
This is very complex and most of it might not be relevant to someone writing a custom connector. I wonder if we can somehow comment which parts of this example are not relevant, or maybe use a better, simpler example.